File: examples/python/core/stream_case.py
# ==========================================================================
# APP v1.1.4
# core/stream_case.py
# --------------------------------------------------------------------------
# Base contract for the stream surface in APP.
#
# Represents event-driven execution.
#
# Responsibility:
# - consume events
# - process business logic (service) or orchestrate (composition)
# - produce new events or side effects
#
# Used for: queues, webhooks, event bus, async pipelines.
#
# Context:
# - StreamContext extends AppBaseContext with event infrastructure
# - each project defines the concrete types for eventBus, queue, etc.
# ==========================================================================

from __future__ import annotations

import math
import random
import traceback
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, Generic, TypeVar

from core.shared.app_base_context import AppBaseContext
from core.shared.app_infra_contracts import AppCache, AppEventPublisher
from core.shared.app_structural_contracts import StreamFailureEnvelope

TInput = TypeVar("TInput")
TOutput = TypeVar("TOutput")


# ==========================================================================
# StreamContext
# ==========================================================================

class StreamContext(AppBaseContext, total=False):
    """Context for the stream surface."""

    # Event publisher (publish side only — consume is in stream surface).
    event_bus: AppEventPublisher

    # Queue access (kept as Any — needs normative note distinguishing from event_bus).
    queue: Any

    # Database access (kept as Any — no stable contract).
    db: Any

    # Cache with optional TTL.
    cache: AppCache

    # Cases loaded by the runtime (for cross-case composition).
    cases: dict[str, Any]

    # Library packages registered by the host.
    packages: dict[str, Any]

    # Free extension space for the host.
    extra: dict[str, Any]


# ==========================================================================
# StreamEvent
# ==========================================================================

@dataclass
class StreamEvent(Generic[TInput]):
    """Generic event structure."""

    type: str
    payload: TInput

    # Idempotency key for at-least-once delivery deduplication.
    idempotency_key: str | None = None

    metadata: dict[str, Any] = field(default_factory=dict)


# ==========================================================================
# AppStreamRecoveryPolicy
# --------------------------------------------------------------------------
# Declarative recovery contract for stream capabilities.
#
# This policy describes intended semantics only.
# The app host is responsible for validating compatibility with the
# chosen runtime and for translating the contract to platform-specific
# configuration.
# ==========================================================================

@dataclass
class AppStreamRetryPolicy:
    """Retry configuration within a recovery policy."""

    # Total number of attempts including the first execution.
    # max_attempts = 1 means fail-fast, no retry.
    max_attempts: int = 1

    backoff_ms: int | None = None
    multiplier: float | None = None
    max_backoff_ms: int | None = None
    jitter: bool = False
    retryable_errors: list[str] | None = None


@dataclass
class AppStreamDeadLetterConfig:
    """Dead-letter configuration within a recovery policy."""

    # Logical dead-letter destination identifier.
    # Must be bound by the app host to a physical transport destination.
    destination: str = ""

    include_failure_metadata: bool = False


@dataclass
class AppStreamRecoveryPolicy:
    """Declarative recovery contract for stream capabilities."""
    retry: AppStreamRetryPolicy | None = None
    dead_letter: AppStreamDeadLetterConfig | None = None


# ==========================================================================
# Recovery policy validation
# --------------------------------------------------------------------------
# Protocol-level shape validation for recovery metadata.
# Host-specific compatibility checks remain the app's responsibility.
# ==========================================================================

def validate_stream_recovery_policy(
    source: str,
    policy: AppStreamRecoveryPolicy | None,
) -> None:
    """Validate canonical APP invariants for a recovery policy."""
    if policy is None:
        return

    label = source or "stream"
    retry = policy.retry
    dead_letter = policy.dead_letter

    if retry is not None:
        if not isinstance(retry.max_attempts, int) or retry.max_attempts < 1:
            raise ValueError(
                f"{label}: recoveryPolicy.retry.max_attempts must be an integer >= 1"
            )
        if retry.backoff_ms is not None and retry.backoff_ms < 0:
            raise ValueError(f"{label}: recoveryPolicy.retry.backoff_ms must be >= 0")
        if retry.multiplier is not None and retry.multiplier < 1:
            raise ValueError(f"{label}: recoveryPolicy.retry.multiplier must be >= 1")
        if retry.max_backoff_ms is not None and retry.max_backoff_ms < 0:
            raise ValueError(f"{label}: recoveryPolicy.retry.max_backoff_ms must be >= 0")
        if (
            retry.backoff_ms is not None
            and retry.max_backoff_ms is not None
            and retry.max_backoff_ms < retry.backoff_ms
        ):
            raise ValueError(
                f"{label}: recoveryPolicy.retry.max_backoff_ms must be >= backoff_ms"
            )
        if retry.retryable_errors is not None and any(
            code.strip() == "" for code in retry.retryable_errors
        ):
            raise ValueError(
                f"{label}: recoveryPolicy.retry.retryable_errors must contain stable non-empty codes"
            )

    if dead_letter is not None and dead_letter.destination.strip() == "":
        raise ValueError(
            f"{label}: recoveryPolicy.dead_letter.destination must be a non-empty logical identifier"
        )


def is_stream_error_retryable(
    error: BaseException,
    retryable_errors: list[str] | None,
) -> bool:
    """Check if an error is retryable given the retryable_errors list."""
    if retryable_errors is None or len(retryable_errors) == 0:
        return True

    code = _extract_stream_error_code(error)
    return code in retryable_errors if code is not None else False


def compute_stream_retry_delay_ms(
    retry: AppStreamRetryPolicy,
    attempt: int,
) -> int:
    """Compute the delay in milliseconds for a retry attempt."""
    base = retry.backoff_ms if retry.backoff_ms is not None else 0
    if base <= 0:
        return 0

    multiplier = retry.multiplier if retry.multiplier is not None else 1.0
    exponent = max(0, attempt - 1)
    delay = base * math.pow(multiplier, exponent)

    if retry.max_backoff_ms is not None:
        delay = min(delay, retry.max_backoff_ms)

    if retry.jitter and delay > 0:
        delay = random.random() * delay

    return int(delay)


def create_stream_failure_envelope(
    case_name: str,
    event: StreamEvent[Any],
    error: BaseException,
    attempts: int,
    correlation_id: str,
    first_attempt_at: str,
    last_attempt_at: str,
) -> StreamFailureEnvelope[StreamEvent[Any]]:
    """Create a dead-letter failure envelope."""
    code = _extract_stream_error_code(error)
    message = str(error) if isinstance(error, Exception) else "Unknown stream failure"
    stack = traceback.format_exception(error) if isinstance(error, Exception) else None

    return StreamFailureEnvelope(
        case_name=case_name,
        surface="stream",
        original_event=event,
        last_error_message=message,
        last_error_code=code,
        last_error_stack="".join(stack) if stack else None,
        attempts=attempts,
        first_attempt_at=first_attempt_at,
        last_attempt_at=last_attempt_at,
        correlation_id=correlation_id,
    )


def _extract_stream_error_code(error: BaseException) -> str | None:
    """Extract a code attribute from an error, if present."""
    code = getattr(error, "code", None)
    return code if isinstance(code, str) else None


# ==========================================================================
# BaseStreamCase
# ==========================================================================

class BaseStreamCase(ABC, Generic[TInput, TOutput]):
    """Base class for event-driven stream surfaces in APP."""

    def __init__(self, ctx: StreamContext) -> None:
        self.ctx = ctx

    # ==================================================================
    # Required methods
    # ==================================================================

    @abstractmethod
    def handler(self, event: StreamEvent[TInput]) -> None:
        """
        Main event handler.

        Receives a business event and processes it.
        Transport bindings (topic subscriptions, queue listeners)
        live in subscribe() or in the adapter/host.
        """
        ...

    def subscribe(self) -> Any:
        """Subscription registration. Declarative binding."""
        return None

    def recovery_policy(self) -> AppStreamRecoveryPolicy | None:
        """
        Declarative recovery contract.

        Must be deterministic, serializable, free of callbacks,
        and independent of the event payload.
        The app host validates and translates this policy to the real runtime.
        """
        return None

    def test(self) -> None:
        """Self-contained stream surface test."""

    # ==================================================================
    # Canonical internal slots
    # ==================================================================

    def _repository(self) -> Any:
        """Persistence and local integrations (idempotency, checkpoints)."""
        return None

    def _composition(self, event: StreamEvent[TInput]) -> None:
        """
        Cross-case orchestration via registry (composed Case).

        When present, pipeline() uses _composition as the main center.
        """

    def _consume(self, event: StreamEvent[TInput]) -> TInput:
        """Initial event consumption."""
        return event.payload

    def _service(self, input: TInput) -> TOutput | None:
        """Atomic business logic for the stream."""
        return None

    def _publish(self, output: TOutput) -> None:
        """Publish a resulting event."""

    # ==================================================================
    # Pipeline
    # ==================================================================

    def pipeline(self, event: StreamEvent[TInput]) -> None:
        """
        Default execution pipeline.

        If _composition is overridden, delegates to it (composed Case).
        Otherwise, orchestrates the atomic flow: consume -> service -> publish.

        The default pipeline does not implement retry, backoff, or dead-letter.
        Recovery is the app host/runtime's responsibility when recovery_policy()
        is declared.
        """
        # Check if _composition was overridden from the base no-op
        if type(self)._composition is not BaseStreamCase._composition:
            self._composition(event)
            return

        consumed = self._consume(event)

        transformed = self._service(consumed)

        if transformed is not None:
            self._publish(transformed)

File: examples/deno/core/stream.case.ts
/* ========================================================================== *
 * APP v1.1.6
 * core/stream.case.ts
 * ----------------------------------------------------------------------------
 * Base contract for the APP stream surface.
 *
 * Represents event-driven execution.
 *
 * Responsibility:
 * - consume events
 * - process business logic (service) or orchestrate (composition)
 * - produce new events or side effects
 *
 * Used for:
 * - queues
 * - webhooks
 * - event bus
 * - asynchronous pipelines
 *
 * Context:
 * - StreamContext extends AppBaseContext with event infrastructure
 * - each project defines the concrete types for eventBus, queue, etc.
 * ========================================================================== */

import { Dict } from "./domain.case";
import { AppBaseContext } from "./shared/app_base_context";
import { StreamFailureEnvelope } from "./shared/app_structural_contracts";
import { AppCache, AppEventPublisher } from "./shared/app_infra_contracts";

/* ==========================================================================
 * StreamContext
 * --------------------------------------------------------------------------
 * Stream surface-specific context.
 *
 * Extends AppBaseContext with event infrastructure:
 * - eventBus: event publisher (publish side only — consume lives in the stream surface)
 * - queue: queue access (unknown — needs a normative note distinguishing it from eventBus)
 * - db: database access (unknown — no stable contract)
 * - cache: cache with TTL
 *
 * Fields with a minimum contract use interfaces from app_infra_contracts.ts.
 * Fields without stable semantics remain unknown.
 * ========================================================================== */

export interface StreamContext extends AppBaseContext {
  /**
   * Event publisher.
   *
   * Covers the publish side of event-driven communication.
   * The consume/subscribe side is modeled by BaseStreamCase.subscribe().
   *
   * Examples: Kafka producer, RabbitMQ channel, Redis Pub/Sub publisher.
   */
  eventBus?: AppEventPublisher;

  /**
   * Queue access.
   *
   * Kept as unknown — it is too close to eventBus and needs
   * a normative note distinguishing it before receiving a minimum contract.
   *
   * Examples: SQS client, BullMQ queue, Cloud Tasks.
   */
  queue?: unknown;

  /**
   * Database access.
   *
   * Kept as unknown — no stable contract.
   * Useful for: idempotency, checkpoints, pipeline state.
   */
  db?: unknown;

  /**
   * Cache with optional TTL.
   *
   * Useful for: deduplication, retry control, processing windows.
   */
  cache?: AppCache;

  /**
   * Registry of Cases loaded by the runtime.
   *
   * Allows cross-case composition through the registry boundary.
   * Used by `_composition` to resolve capabilities from other Cases.
   *
   * Example: ctx.cases?.users?.user_validate?.api?.handler(input)
   */
  cases?: Dict;

  /**
   * Library packages registered by the host.
   *
   * Exposed through registry._packages.
   * Pure libraries from packages/ that the app makes available.
   */
  packages?: Dict;

  /**
   * Free extension space for the project host.
   */
  extra?: Dict;
}

/* ==========================================================================
 * StreamEvent
 * ========================================================================== */

/**
 * Generic event structure.
 */
export interface StreamEvent<T = unknown> {
  type: string;
  payload: T;

  /**
   * Event idempotency key.
   *
   * Allows consumers to detect and discard duplicate events.
   * Essential for production with brokers that guarantee at-least-once
   * delivery (SQS, Kafka, RabbitMQ, EventBridge).
   *
   * The protocol does not dictate the format — it may be a UUID, payload hash,
   * or composed business key. Responsibility for generating and
   * verifying the key belongs to the producer and consumer respectively.
   */
  idempotencyKey?: string;

  metadata?: Record<string, unknown>;
}

/* ==========================================================================
 * AppStreamRecoveryPolicy
 * --------------------------------------------------------------------------
 * Declarative recovery contract for stream capabilities.
 *
 * This policy describes intended semantics only.
 * The app host is responsible for validating compatibility with the
 * chosen runtime and for translating the contract to platform-specific
 * configuration.
 * ========================================================================== */

export interface AppStreamRecoveryPolicy {
  retry?: {
    /**
     * Total number of attempts, including the first execution.
     *
     * maxAttempts: 1 = fail-fast, no retry.
     */
    maxAttempts: number;
    backoffMs?: number;
    multiplier?: number;
    maxBackoffMs?: number;
    jitter?: boolean;
    retryableErrors?: string[];
  };

  deadLetter?: {
    /**
     * Logical dead-letter destination identifier.
     *
     * Must be bound by the app host to a physical transport destination.
     */
    destination: string;
    includeFailureMetadata?: boolean;
  };
}

export interface AppStreamDeadLetterBinding<TEvent = unknown> {
  publish(envelope: StreamFailureEnvelope<StreamEvent<TEvent>>): Promise<void>;
}

export interface AppStreamRuntimeCapabilities {
  maxAttemptsLimit?: number;
  supportsJitter?: boolean;
  deadLetters?: Dict<AppStreamDeadLetterBinding>;
}

/* ==========================================================================
 * Policy validation
 * --------------------------------------------------------------------------
 * Protocol-level shape validation for recovery metadata.
 *
 * This validates the canonical APP invariants that are independent of any
 * specific runtime. Host-specific compatibility checks (DLQ bindings,
 * jitter support, attempt limits, etc.) remain the app's responsibility.
 * ========================================================================== */

export function validateStreamRecoveryPolicy(
  source: string,
  policy?: AppStreamRecoveryPolicy,
): void {
  if (!policy) return;

  const label = source || "stream";
  const retry = policy.retry;
  const deadLetter = policy.deadLetter;

  if (retry) {
    if (!Number.isInteger(retry.maxAttempts) || retry.maxAttempts < 1) {
      throw new Error(
        `${label}: recoveryPolicy.retry.maxAttempts must be an integer >= 1`,
      );
    }

    if (retry.backoffMs !== undefined && retry.backoffMs < 0) {
      throw new Error(`${label}: recoveryPolicy.retry.backoffMs must be >= 0`);
    }

    if (retry.multiplier !== undefined && retry.multiplier < 1) {
      throw new Error(`${label}: recoveryPolicy.retry.multiplier must be >= 1`);
    }

    if (retry.maxBackoffMs !== undefined && retry.maxBackoffMs < 0) {
      throw new Error(
        `${label}: recoveryPolicy.retry.maxBackoffMs must be >= 0`,
      );
    }

    if (
      retry.backoffMs !== undefined &&
      retry.maxBackoffMs !== undefined &&
      retry.maxBackoffMs < retry.backoffMs
    ) {
      throw new Error(
        `${label}: recoveryPolicy.retry.maxBackoffMs must be >= backoffMs`,
      );
    }

    if (
      retry.retryableErrors?.some((code) => code.trim().length === 0)
    ) {
      throw new Error(
        `${label}: recoveryPolicy.retry.retryableErrors must contain stable non-empty codes`,
      );
    }
  }

  if (deadLetter && deadLetter.destination.trim().length === 0) {
    throw new Error(
      `${label}: recoveryPolicy.deadLetter.destination must be a non-empty logical identifier`,
    );
  }
}

export function validateStreamRuntimeCompatibility(
  source: string,
  policy: AppStreamRecoveryPolicy | undefined,
  runtime: AppStreamRuntimeCapabilities,
): void {
  if (!policy) return;

  const label = source || "stream";
  const retry = policy.retry;
  const deadLetter = policy.deadLetter;

  if (
    retry?.maxAttempts !== undefined &&
    runtime.maxAttemptsLimit !== undefined &&
    retry.maxAttempts > runtime.maxAttemptsLimit
  ) {
    throw new Error(
      `${label}: recoveryPolicy.retry.maxAttempts=${retry.maxAttempts} exceeds host limit ${runtime.maxAttemptsLimit}`,
    );
  }

  if (retry?.jitter && runtime.supportsJitter === false) {
    throw new Error(
      `${label}: recoveryPolicy.retry.jitter=true but host runtime does not support jitter`,
    );
  }

  if (
    deadLetter &&
    !runtime.deadLetters?.[deadLetter.destination]
  ) {
    throw new Error(
      `${label}: dead-letter destination "${deadLetter.destination}" is not bound by the host app`,
    );
  }
}

export function isStreamErrorRetryable(
  error: unknown,
  retryableErrors?: string[],
): boolean {
  if (!retryableErrors || retryableErrors.length === 0) {
    return true;
  }

  const code = extractStreamErrorCode(error);
  return code ? retryableErrors.includes(code) : false;
}

export function computeStreamRetryDelayMs(
  retry: NonNullable<AppStreamRecoveryPolicy["retry"]>,
  attempt: number,
): number {
  const base = retry.backoffMs ?? 0;
  if (base <= 0) return 0;

  const multiplier = retry.multiplier ?? 1;
  const exponent = Math.max(0, attempt - 1);
  let delay = base * Math.pow(multiplier, exponent);

  if (retry.maxBackoffMs !== undefined) {
    delay = Math.min(delay, retry.maxBackoffMs);
  }

  if (retry.jitter && delay > 0) {
    delay = Math.floor(Math.random() * delay);
  }

  return Math.floor(delay);
}

export function createStreamFailureEnvelope<TEvent>(
  caseName: string,
  event: StreamEvent<TEvent>,
  error: unknown,
  attempts: number,
  correlationId: string,
  firstAttemptAt: string,
  lastAttemptAt: string,
): StreamFailureEnvelope<StreamEvent<TEvent>> {
  const code = extractStreamErrorCode(error);
  const message = error instanceof Error
    ? error.message
    : "Unknown stream failure";
  const stack = error instanceof Error ? error.stack : undefined;

  return {
    caseName,
    surface: "stream",
    originalEvent: event,
    lastError: {
      message,
      ...(code !== undefined && { code }),
      ...(stack !== undefined && { stack }),
    },
    attempts,
    firstAttemptAt,
    lastAttemptAt,
    correlationId,
  };
}

function extractStreamErrorCode(error: unknown): string | undefined {
  if (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    typeof (error as { code?: unknown }).code === "string"
  ) {
    return (error as { code: string }).code;
  }

  return undefined;
}

/* ==========================================================================
 * BaseStreamCase
 * ========================================================================== */

/**
 * Base class for event-driven execution.
 */
export abstract class BaseStreamCase<TInput = unknown, TOutput = unknown> {
  protected readonly ctx: StreamContext;

  constructor(ctx: StreamContext) {
    this.ctx = ctx;
  }

  /* =======================================================================
   * Required methods
   * ===================================================================== */

  /**
   * Primary event handler.
   *
   * handler is the public entrypoint of the stream capability.
   * It receives a business event and processes it.
   *
   * Transport bindings (topic subscriptions, queue listeners)
   * live in subscribe() or in the adapter/host.
   */
  public abstract handler(event: StreamEvent<TInput>): Promise<void>;

  /**
   * Subscription registration.
   */
  public subscribe?(): unknown;

  /**
   * Contractual recovery declaration.
   *
   * The return value must be pure metadata:
   * - deterministic
   * - serializable
   * - without callbacks
   * - independent from the event payload
   *
   * The app host validates and translates this policy to the real runtime.
   */
  public recoveryPolicy?(): AppStreamRecoveryPolicy;

  /**
   * Internal capability test.
   *
   * Recommended APP practice — surfaces should ideally expose a
   * test() method for self-contained contract validation.
   *
   * Canonical signature: test(): Promise<void>
   * The test creates events internally and invokes handler()/pipeline().
   * It receives no input — it is internal validation.
   */
  public async test(): Promise<void> {}

  /* =======================================================================
   * Internal canonical slots
   * ===================================================================== */

  /**
   * Access to persistence and local integrations for the Case.
   *
   * Canonical slot for idempotency, checkpoints, and pipeline state.
   *
   * Rule: _repository must not perform cross-case composition.
   */
  protected _repository?(): unknown;

  /**
   * Cross-case orchestration via the registry (composed Case).
   *
   * Canonical slot for stream Cases that need to invoke other Cases.
   * Resolves capabilities through ctx.cases, never through direct imports.
   *
   * When present, the pipeline must use _composition as the main center.
   */
  protected async _composition?(event: StreamEvent<TInput>): Promise<void>;

  /* =======================================================================
   * Internal hooks
   * ===================================================================== */

  /**
   * Initial event consumption.
   */
  protected async _consume?(event: StreamEvent<TInput>): Promise<TInput>;

  /**
   * Atomic stream business logic (atomic Case).
   *
   * Canonical slot for processing the consumed event.
   * It receives input, processes it, and produces output.
   *
   * Mutually exclusive with _composition as the main execution center.
   */
  protected async _service?(input: TInput): Promise<TOutput>;

  /**
   * Resulting event publication.
   */
  protected async _publish?(output: TOutput): Promise<void>;

  /**
   * Standard execution pipeline.
   *
   * If _composition is defined, it delegates to it (composed Case).
   * Otherwise, it orchestrates the atomic flow: consume → service → publish.
   *
   * The default pipeline does not implement retry, backoff, or dead-letter.
   * Recovery is the responsibility of the app host/runtime when recoveryPolicy()
   * is declared.
   */
  protected async pipeline(event: StreamEvent<TInput>): Promise<void> {
    if (this._composition) {
      await this._composition(event);
      return;
    }

    const consumed = this._consume ? await this._consume(event) : event.payload;

    const transformed = this._service
      ? await this._service(consumed)
      : (consumed as unknown as TOutput);

    if (this._publish) {
      await this._publish(transformed);
    }
  }
}
